### DISCUSSION TITLE
Semester 4 Woche 6

### DISCUSSION THREAD

**Discussion by TimSchnur** [Christian-2003/CrInGE#47] [Student Exchange]
Hallo zusammen,

diese Woche haben wir uns mit einigen 📊 Software Metriken und deren Analyse beschäftigt.
Nach intensiver Suche haben wir uns schlussendlich für die Folgenden entschieden:

### 1. Linting

Nachdem wir uns diese Woche nochmals mit dem Thema ♻️ Refactoring auseinander gesetzt haben, sind uns einige Unregelmäßigkeiten in den neuen Codestellen aufgefallen. Um das schnelle 💭 Verständnis des Codes zu steigern, wollen wir wieder eine einheitliche und klare äußere Struktur anstreben. Hierfür benötigt es nun einer wohl definierte Formatierung und Code Styles.

Deshalb haben wir das Maven Plugin Checkstyle in das Projekt eingebunden. Hierdurch können Reports erstellt werden, welche 💡 Tipps, ⚠️ Warnungen und ❌ Fehlern bezüglichen der äußeren Form auflisten.

Da wir hierdurch nun einige neuen Linting Regeln haben, liegt unser momentaner Messwert bei **21 Fehlern pro 100 Zeilen** Code.
Diesen wollen wir in den kommenden Wochen 📉 reduzieren und zukünftig möglichst gering halten.

### 2. Zyklomatische Komplexität

Da für das 💭 Verständnis des Codes nicht die äußere Form sondern die logische Komplexität maßgeblich ist, wollen wir zusätzlich die 🔀 zyklomatische Komplexität der Klassen betrachten. Diese spiegelt die Anzahl der möglichen 🔱 Ablaufpfade wieder.

Diese werden wir mittels eines IntelliJ Plugins betrachten. Dieses zeigt direkt zu jeder Klasse die zyklomatische Komplexität an und ermöglicht uns somit das 📋 Erheben von Messwerten.

Darstellen wollen wir die zyklomatische Komplexität als Durchschnitt pro Klasse.
Momentan liegt diese mit **11.6** am unteren Ende des 🟡 mittleren Bereiches und sollte zukünftig 🔴 20 nicht überschreiten.

### 3. Dependencies

Zuletzt haben wir die 🔢 Anzahl der Dependencies gewählt, da der Sinn und unsere oberste Anforderung an die Software ist mit möglichst keinen Abhängigkeiten zu arbeiten.

Momentan nutzen wir **ausschließlich OpenGL** als unsere einzige 📚 Fremdbibliothek.

Das wäre dann auch schon alles für diese Woche. Im nächsten Blog werden wir euch weiter über das Projekt informieren.

**Comment by YPuest** [comment]
Hey,

eure Metriken wirken sehr durchdacht und gut gewählt. Mich persönlich würde noch interessieren, welches Plugin ihr zum Messen der _zyklischen Komplexität_ nutzt.

Da ich euren Blog heute zum ersten mal sehe möchte ich auch noch anmerken, dass mir die äußere Form sehr gut gefällt. Man sieht deutlich, dass ihr euch Mühe gebt. Ich freue mich schon darauf, wenn ich das nächste Mal bei euch vorbei schaue :D

**Reply by Christian-2003** [reply]
Hi,

es freut uns zu hören, dass dir die Form unseres Blogs gefällt. Vielleicht schaust du dann ja in den nächsten Wochen öfters bei uns vorbei, was uns sehr freuen würde.

Um zu deiner Frage zu kommen: Wir verwenden zum Messen der zyklomatischen Komplexität das Plugin **CodeMetrics** für die IntelliJ IDEA. Damit werden die berechneten Werte direkt im Klassen- oder Methodenkopf im Quellcode angezeigt. Dadurch bemerken wir direkt bei der Entwicklung, wenn die entsprechenden Werte zu hoch sind.

**Comment by davidlzt** [comment]
Hey,
die gewählten Metriken scheinen alle Sinnvoll und sehr ausgearbeitet zu sein. Finde es auch gut das eure oberste Anforderung an die Software ist mit möglichst keinen Abhängigkeiten zu arbeiten. Gute Arbeit.
Freue mich auf die nächste Woche :D
